pkg/vsphere/performance/host_metrics.go

```diff
@@ -64,17 +63,41 @@ type HostMetricsInfo struct {
 	CPU    HostCPU
 }
 
-// HostMetrics returns CPU and memory metrics for all ESXi hosts in the input session's cluster.
-func HostMetrics(op trace.Operation, session *session.Session) (map[*object.HostSystem]*HostMetricsInfo, error) {
-	if session == nil {
+// HostMetrics returns CPU and memory metrics for all ESXi hosts in a cluster
+// via implementation of the MetricsProvider interface.
+type HostMetricsProvider struct {
+	Session *session.Session
+}
+
+func NewHostMetricsProvider(s *session.Session) *HostMetricsProvider {
+	return &HostMetricsProvider{Session: s}
+}
+
+// GatherMetricsForComputeResource gathers host metrics from the supplied compute resource.
+func (h *HostMetricsProvider) GetMetricsForComputeResource(op trace.Operation, cr *object.ComputeResource) (map[*object.HostSystem]*HostMetricsInfo, error) {
+	if h.Session == nil {
 		return nil, fmt.Errorf("session not set")
 	}
 
 	// Gather hosts from the session cluster and then obtain their morefs.
-	hosts, err := gatherHosts(op.Context, session)
+	hosts, err := cr.Hosts(op)
 	if err != nil {
-		return nil, fmt.Errorf("unable to obtain host morefs from session: %s", err)
+		return nil, fmt.Errorf("unable to obtain host morefs from compute resource: %s", err)
+	}
+
+	if hosts == nil {
+		return nil, fmt.Errorf("no hosts found in compute resource")
+	}
+
+	return h.GetMetricsForHosts(op, hosts)
+}
+
+// GetMetricsForHosts returns metrics pertaining to supplied ESX hosts.
+func (h *HostMetricsProvider) GetMetricsForHosts(op trace.Operation, hosts []*object.HostSystem) (map[*object.HostSystem]*HostMetricsInfo, error) {
```

Comment:
Since this func is exported and `hosts` is now a parameter, we should add a `len(hosts) == 0` check and bail early if true.

Reply:
👍 fixed